File: js/test/regression-170.js
/**
 * Regression tests for issue #170: detached Docker executions never persist a
 * terminal state, and `--status` fabricates `endTime` with `new Date()`.
 *
 * The three defects covered here:
 *   170.1 the detached completion watcher must write the terminal state back
 *         into the store instead of leaving the record `executing` forever;
 *   170.2 `--status` must derive `endTime` from a real clock
 *         (`State.FinishedAt` -> log footer `Finished:`) and mark the
 *         provenance via `endTimeSource` when it falls back to observation time;
 *   170.3 `cleanupStale()` must record `staleDetectedAt`, not a fabricated
 *         finish time.
 */

const { describe, it, expect, afterEach } = require('bun:test');
const fs = require('fs');
const path = require('path');
const os = require('os');

const {
  ExecutionStore,
  ExecutionRecord,
  ExecutionStatus,
} = require('../src/lib/execution-store');
const { resolveDetachedStatus } = require('../src/lib/status-formatter');
const {
  buildDetachedDockerCompletionScript,
} = require('../src/lib/docker-cleanup');
const { finalizeDetachedExecution } = require('../src/lib/detached-finalize');

const tempDirs = [];

function makeTempDir(prefix) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), prefix));
  tempDirs.push(dir);
  return dir;
}

function createExecutable(filePath, content) {
  fs.writeFileSync(filePath, content, 'utf8');
  fs.chmodSync(filePath, 0o755);
}

/**
 * A fake `docker` that answers `inspect` with a fixed state line, whatever the
 * `-f` template asks for. The template is echoed verbatim by the caller, so the
 * test controls exactly how many fields the formatter has to cope with.
 */
function withFakeDockerInspect(stateLine, fn) {
  const fakeBin = makeTempDir('fake-docker-170-');
  const dockerPath = path.join(fakeBin, 'docker');
  createExecutable(
    dockerPath,
    [
      '#!/bin/sh',
      '[ "$1" = "inspect" ] || exit 1',
      `echo "${stateLine}"`,
      '',
    ].join('\n')
  );
  const originalPath = process.env.PATH;
  const originalDockerBin = process.env.START_DOCKER_BIN;
  process.env.PATH = `${fakeBin}${path.delimiter}${originalPath || ''}`;
  process.env.START_DOCKER_BIN = dockerPath;
  try {
    return fn();
  } finally {
    process.env.PATH = originalPath;
    if (originalDockerBin === undefined) {
      delete process.env.START_DOCKER_BIN;
    } else {
      process.env.START_DOCKER_BIN = originalDockerBin;
    }
  }
}

function withFakeDockerMissingContainer(fn) {
  const fakeBin = makeTempDir('fake-docker-gone-170-');
  const dockerPath = path.join(fakeBin, 'docker');
  createExecutable(dockerPath, ['#!/bin/sh', 'exit 1', ''].join('\n'));
  const originalPath = process.env.PATH;
  const originalDockerBin = process.env.START_DOCKER_BIN;
  process.env.PATH = `${fakeBin}${path.delimiter}${originalPath || ''}`;
  process.env.START_DOCKER_BIN = dockerPath;
  try {
    return fn();
  } finally {
    process.env.PATH = originalPath;
    if (originalDockerBin === undefined) {
      delete process.env.START_DOCKER_BIN;
    } else {
      process.env.START_DOCKER_BIN = originalDockerBin;
    }
  }
}

function detachedDockerRecord(overrides = {}) {
  return new ExecutionRecord({
    pid: null,
    status: ExecutionStatus.EXECUTING,
    exitCode: null,
    command: 'sleep 100',
    startTime: '2026-09-15T22:21:40.000Z',
    endTime: null,
    options: {
      isolated: 'docker',
      isolationMode: 'detached',
      sessionName: 'start-command-170',
    },
    ...overrides,
  });
}

afterEach(() => {
  while (tempDirs.length > 0) {
    const dir = tempDirs.pop();
    fs.rmSync(dir, { recursive: true, force: true });
  }
});

describe('issue #170.2: endTime is never fabricated silently', () => {
  it('derives endTime from docker State.FinishedAt for a stopped container', () => {
    const record = detachedDockerRecord();
    const enriched = withFakeDockerInspect(
      'false 137 false 2026-09-15T22:21:40.942007645Z 2026-09-15T22:21:46.740817278Z',
      () => resolveDetachedStatus(record, null)
    );

    expect(enriched.status).toBe('executed');
    expect(enriched.exitCode).toBe(137);
    expect(enriched.endTimeSource).toBe('docker-finished-at');
    expect(new Date(enriched.endTime).toISOString()).toBe(
      '2026-09-15T22:21:46.740Z'
    );
  });

  it('falls back to the log footer Finished: line when docker is gone', () => {
    const record = detachedDockerRecord();
    const tail = [
      'output line',
      '='.repeat(50),
      'Finished: 2026-09-15 22:21:46.740',
      'Exit Code: 137',
      '',
    ].join('\n');

    const enriched = withFakeDockerMissingContainer(() =>
      resolveDetachedStatus(record, tail)
    );

    expect(enriched.status).toBe('executed');
    expect(enriched.exitCode).toBe(137);
    expect(enriched.endTimeSource).toBe('log-footer');
    expect(new Date(enriched.endTime).toISOString()).toBe(
      '2026-09-15T22:21:46.740Z'
    );
  });

  it('marks the observation-time fallback instead of pretending it is a finish time', () => {
    const record = detachedDockerRecord();
    const tail = [
      '='.repeat(50),
      'Finished: not-a-timestamp',
      'Exit Code: 3',
      '',
    ].join('\n');

    const enriched = withFakeDockerMissingContainer(() =>
      resolveDetachedStatus(record, tail)
    );

    expect(enriched.status).toBe('executed');
    expect(enriched.exitCode).toBe(3);
    expect(enriched.endTimeSource).toBe('observed-at');
    expect(enriched.observedAt).toBe(enriched.endTime);
  });

  it('keeps an already recorded endTime and marks it as persisted', () => {
    const record = detachedDockerRecord({
      status: ExecutionStatus.EXECUTED,
      exitCode: 0,
      endTime: '2026-09-15T22:21:46.740Z',
      endTimeSource: 'docker-finished-at',
    });

    const enriched = withFakeDockerInspect('false 0 false', () =>
      resolveDetachedStatus(record, null)
    );

    expect(enriched.endTime).toBe('2026-09-15T22:21:46.740Z');
    expect(enriched.endTimeSource).toBe('docker-finished-at');
  });
});

describe('issue #170.3: cleanupStale records the detection time', () => {
  it('sets staleDetectedAt and leaves endTime unknown', () => {
    const appFolder = makeTempDir('stale-170-');
    const store = new ExecutionStore({ appFolder, useLinks: false });

    const record = new ExecutionRecord({
      pid: 999999,
      status: ExecutionStatus.EXECUTING,
      command: 'sleep 1000',
      startTime: new Date(Date.now() - 48 * 60 * 60 * 1000).toISOString(),
    });
    store.save(record);

    const result = store.cleanupStale({ maxAge: 1000 });
    expect(result.cleaned).toBe(1);

    const cleaned = store.get(record.uuid);
    expect(cleaned.status).toBe(ExecutionStatus.EXECUTED);
    expect(cleaned.exitCode).toBe(-1);
    expect(cleaned.endTime).toBeNull();
    expect(typeof cleaned.staleDetectedAt).toBe('string');
    expect(Number.isNaN(Date.parse(cleaned.staleDetectedAt))).toBe(false);
  });
});

describe('issue #170.1: the detached watcher persists the terminal state', () => {
  it('invokes the finalizer with the inspected docker facts', () => {
    const script = buildDetachedDockerCompletionScript(
      'start-command-170',
      'default',
      '/tmp/start-command-170.log',
      'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee'
    );

    expect(script).toContain('detached-finalize');
    expect(script).toContain('aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee');
    expect(script).toContain('$__start_command_exit');
    expect(script).toContain('$__start_command_finished');
  });

  it('stays backward compatible when no execution id is supplied', () => {
    const script = buildDetachedDockerCompletionScript(
      'start-command-170',
      'default',
      '/tmp/start-command-170.log'
    );
    expect(script).not.toContain('detached-finalize');
  });

  it('writes status, exitCode, endTime and provenance into the store', () => {
    const appFolder = makeTempDir('finalize-170-');
    const store = new ExecutionStore({ appFolder, useLinks: false });
    const record = detachedDockerRecord();
    store.save(record);

    const outcome = finalizeDetachedExecution({
      store,
      executionId: record.uuid,
      exitCode: 137,
      oomKilled: false,
      startedAt: '2026-09-15T22:21:40.942007645Z',
      finishedAt: '2026-09-15T22:21:46.740817278Z',
      containerError: '',
    });

    expect(outcome.updated).toBe(true);

    const stored = store.get(record.uuid);
    expect(stored.status).toBe(ExecutionStatus.EXECUTED);
    expect(stored.exitCode).toBe(137);
    expect(stored.oomKilled).toBe(false);
    expect(stored.endTimeSource).toBe('docker-finished-at');
    expect(new Date(stored.endTime).toISOString()).toBe(
      '2026-09-15T22:21:46.740Z'
    );
    expect(stored.exitReason).toContain('SIGKILL');
  });

  it('falls back to observation time when docker reports no finish time', () => {
    const appFolder = makeTempDir('finalize-170-zero-');
    const store = new ExecutionStore({ appFolder, useLinks: false });
    const record = detachedDockerRecord();
    store.save(record);

    finalizeDetachedExecution({
      store,
      executionId: record.uuid,
      exitCode: 0,
      oomKilled: false,
      startedAt: 'unknown',
      finishedAt: '0001-01-01T00:00:00Z',
      containerError: '',
    });

    const stored = store.get(record.uuid);
    expect(stored.status).toBe(ExecutionStatus.EXECUTED);
    expect(stored.exitCode).toBe(0);
    expect(stored.endTimeSource).toBe('observed-at');
    expect(stored.observedAt).toBe(stored.endTime);
    expect(new Date(stored.endTime).getUTCFullYear()).toBeGreaterThan(2000);
  });

  it('never fails when the record is gone', () => {
    const appFolder = makeTempDir('finalize-170-missing-');
    const store = new ExecutionStore({ appFolder, useLinks: false });
    const outcome = finalizeDetachedExecution({
      store,
      executionId: 'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee',
      exitCode: 1,
    });
    expect(outcome.updated).toBe(false);
    expect(outcome.reason).toBe('record-not-found');
  });
});

File: rust/tests/regression_170.rs
//! Regression tests for issue #170: detached Docker executions never persist a
//! terminal state, and `--status` fabricates `end_time` with the current clock.
//!
//! The three defects covered here:
//!   170.1 the detached completion watcher must write the terminal state back
//!         into the store instead of leaving the record `executing` forever;
//!   170.2 `--status` must derive `end_time` from a real clock
//!         (`State.FinishedAt` -> log footer `Finished:`) and mark the
//!         provenance via `end_time_source` when it falls back to observation
//!         time;
//!   170.3 `cleanup_stale()` must record `stale_detected_at`, not a fabricated
//!         finish time.

use chrono::{DateTime, Datelike, Utc};
use serde_json::Value;
use start_command::{
    enrich_detached_status, finalize_detached_execution, parse_footer_timestamp, CleanupOptions,
    DetachedFinalizeFacts, ExecutionRecord, ExecutionRecordOptions, ExecutionStatus,
    ExecutionStore, ExecutionStoreOptions,
};
use std::collections::HashMap;
use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
use std::path::{Path, PathBuf};
use std::sync::{Mutex, OnceLock};
use tempfile::TempDir;

const CONTAINER_STARTED_AT: &str = "2026-09-15T22:21:40.942007645Z";
const CONTAINER_FINISHED_AT: &str = "2026-09-15T22:21:46.740817278Z";

fn detached_docker_record(log_path: &Path) -> ExecutionRecord {
    let mut options = HashMap::new();
    options.insert(
        "sessionName".to_string(),
        Value::String("start-command-170".to_string()),
    );
    options.insert("isolated".to_string(), Value::String("docker".to_string()));
    options.insert(
        "isolationMode".to_string(),
        Value::String("detached".to_string()),
    );

    ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 100".to_string(),
        uuid: Some("issue170-rust".to_string()),
        status: Some(ExecutionStatus::Executing),
        log_path: Some(log_path.to_string_lossy().to_string()),
        start_time: Some("2026-09-15T22:21:40.000Z".to_string()),
        options: Some(options),
        ..Default::default()
    })
}

fn write_log_with_footer(dir: &Path, finished: &str, exit_code: i32) -> PathBuf {
    let log_path = dir.join("issue170.log");
    let content = format!(
        "output line\n{}\nFinished: {}\nExit Code: {}\n",
        "=".repeat(50),
        finished,
        exit_code
    );
    std::fs::write(&log_path, content).unwrap();
    log_path
}

/// A fake `docker` that answers every `inspect` with a fixed state line,
/// whatever the `-f` template asks for.
fn write_fake_docker(fake_dir: &Path, state_line: &str) -> PathBuf {
    write_docker_script(
        fake_dir,
        &[
            "#!/bin/sh".to_string(),
            "[ \"$1\" = \"inspect\" ] || exit 1".to_string(),
            format!("echo '{}'", state_line),
            String::new(),
        ]
        .join("\n"),
    )
}

/// A fake `docker` whose every `inspect` fails: the container is gone.
fn write_missing_container_docker(fake_dir: &Path) -> PathBuf {
    write_docker_script(fake_dir, "#!/bin/sh\nexit 1\n")
}

#[cfg(not(windows))]
fn write_docker_script(fake_dir: &Path, script: &str) -> PathBuf {
    use std::os::unix::fs::PermissionsExt;

    let docker_path = fake_dir.join("docker");
    std::fs::write(&docker_path, script).unwrap();
    let mut permissions = std::fs::metadata(&docker_path).unwrap().permissions();
    permissions.set_mode(0o755);
    std::fs::set_permissions(&docker_path, permissions).unwrap();
    docker_path
}

#[cfg(windows)]
fn write_docker_script(fake_dir: &Path, script: &str) -> PathBuf {
    // The POSIX bodies above are only ever used on unix hosts; on Windows the
    // shipped watcher shell does not run either, so a stub that always fails is
    // enough to keep the "container is gone" tests meaningful.
    let _ = script;
    let docker_path = fake_dir.join("docker.cmd");
    std::fs::write(&docker_path, "@echo off\r\nexit /b 1\r\n").unwrap();
    docker_path
}

fn with_fake_docker<F: FnOnce()>(write_docker: impl FnOnce(&Path) -> PathBuf, run: F) {
    static FAKE_DOCKER_ENV_LOCK: OnceLock<Mutex<()>> = OnceLock::new();
    let _guard = FAKE_DOCKER_ENV_LOCK
        .get_or_init(|| Mutex::new(()))
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    let fake_dir = TempDir::new().unwrap();
    let docker_path = write_docker(fake_dir.path());
    let original_docker_bin = std::env::var_os("START_DOCKER_BIN");
    std::env::set_var("START_DOCKER_BIN", &docker_path);
    let result = catch_unwind(AssertUnwindSafe(run));
    if let Some(path) = original_docker_bin {
        std::env::set_var("START_DOCKER_BIN", path);
    } else {
        std::env::remove_var("START_DOCKER_BIN");
    }
    if let Err(payload) = result {
        resume_unwind(payload);
    }
}

fn test_store(dir: &Path) -> ExecutionStore {
    ExecutionStore::with_options(ExecutionStoreOptions {
        app_folder: Some(dir.to_path_buf()),
        use_links: Some(false),
        verbose: false,
    })
}

// ---------------------------------------------------------------------------
// 170.2: end_time is never fabricated silently
// ---------------------------------------------------------------------------

#[test]
fn derives_end_time_from_docker_finished_at_for_a_stopped_container() {
    let temp = TempDir::new().unwrap();
    let log_path = write_log_with_footer(temp.path(), "2026-09-15 22:21:46.740", 137);
    let record = detached_docker_record(&log_path);

    with_fake_docker(
        |dir| {
            write_fake_docker(
                dir,
                &format!("false 137 false {CONTAINER_STARTED_AT} {CONTAINER_FINISHED_AT}"),
            )
        },
        || {
            let enriched = enrich_detached_status(&record);
            assert_eq!(enriched.status, ExecutionStatus::Executed);
            assert_eq!(enriched.exit_code, Some(137));
            assert_eq!(
                enriched.end_time_source.as_deref(),
                Some("docker-finished-at")
            );
            let end_time = DateTime::parse_from_rfc3339(enriched.end_time.as_deref().unwrap())
                .unwrap()
                .with_timezone(&Utc);
            assert_eq!(
                end_time.to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
                "2026-09-15T22:21:46.740Z"
            );
            assert_eq!(
                enriched.container_started_at.as_deref(),
                Some(CONTAINER_STARTED_AT)
            );
        },
    );
}

#[test]
fn falls_back_to_the_log_footer_finished_line_when_docker_is_gone() {
    let temp = TempDir::new().unwrap();
    let log_path = write_log_with_footer(temp.path(), "2026-09-15 22:21:46.740", 137);
    let record = detached_docker_record(&log_path);

    with_fake_docker(write_missing_container_docker, || {
        let enriched = enrich_detached_status(&record);
        assert_eq!(enriched.status, ExecutionStatus::Executed);
        assert_eq!(enriched.exit_code, Some(137));
        assert_eq!(enriched.end_time_source.as_deref(), Some("log-footer"));
        let end_time = DateTime::parse_from_rfc3339(enriched.end_time.as_deref().unwrap())
            .unwrap()
            .with_timezone(&Utc);
        assert_eq!(
            end_time.to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
            "2026-09-15T22:21:46.740Z"
        );
    });
}

#[test]
fn marks_the_observation_time_fallback_instead_of_pretending_it_is_a_finish_time() {
    let temp = TempDir::new().unwrap();
    let log_path = write_log_with_footer(temp.path(), "not-a-timestamp", 3);
    let record = detached_docker_record(&log_path);

    with_fake_docker(write_missing_container_docker, || {
        let enriched = enrich_detached_status(&record);
        assert_eq!(enriched.status, ExecutionStatus::Executed);
        assert_eq!(enriched.exit_code, Some(3));
        assert_eq!(enriched.end_time_source.as_deref(), Some("observed-at"));
        assert_eq!(enriched.observed_at, enriched.end_time);
    });
}

#[test]
fn keeps_an_already_recorded_end_time_and_its_provenance() {
    let temp = TempDir::new().unwrap();
    let log_path = write_log_with_footer(temp.path(), "2026-09-15 22:21:46.740", 0);
    let mut record = detached_docker_record(&log_path);
    record.status = ExecutionStatus::Executed;
    record.exit_code = Some(0);
    record.end_time = Some("2026-09-15T22:21:46.740Z".to_string());
    record.end_time_source = Some("docker-finished-at".to_string());

    with_fake_docker(
        |dir| {
            write_fake_docker(
                dir,
                "false 0 false 0001-01-01T00:00:00Z 0001-01-01T00:00:00Z",
            )
        },
        || {
            let enriched = enrich_detached_status(&record);
            assert_eq!(
                enriched.end_time.as_deref(),
                Some("2026-09-15T22:21:46.740Z")
            );
            assert_eq!(
                enriched.end_time_source.as_deref(),
                Some("docker-finished-at")
            );
        },
    );
}

#[test]
fn rejects_footer_timestamps_that_are_not_real_timestamps() {
    assert!(parse_footer_timestamp("not-a-timestamp").is_none());
    assert!(parse_footer_timestamp("").is_none());
    assert_eq!(
        parse_footer_timestamp("2026-09-15 22:21:46.740").as_deref(),
        Some("2026-09-15T22:21:46.740+00:00")
    );
}

// ---------------------------------------------------------------------------
// 170.3: cleanup_stale records the detection time
// ---------------------------------------------------------------------------

#[test]
fn cleanup_stale_sets_stale_detected_at_and_leaves_end_time_unknown() {
    let temp = TempDir::new().unwrap();
    let store = test_store(temp.path());

    let mut record = ExecutionRecord::new("sleep 1000");
    record.start_time = (Utc::now() - chrono::Duration::hours(48)).to_rfc3339();
    store.save(&record).unwrap();

    let result = store.cleanup_stale(CleanupOptions {
        dry_run: false,
        ..Default::default()
    });
    assert_eq!(result.cleaned, 1);

    let cleaned = store.get(&record.uuid).unwrap();
    assert_eq!(cleaned.status, ExecutionStatus::Executed);
    assert_eq!(cleaned.exit_code, Some(-1));
    assert!(cleaned.end_time.is_none());
    let detected = cleaned.stale_detected_at.expect("stale_detected_at");
    assert!(DateTime::parse_from_rfc3339(&detected).is_ok());
}

// ---------------------------------------------------------------------------
// 170.1: the detached watcher persists the terminal state
// ---------------------------------------------------------------------------

#[test]
fn finalize_writes_status_exit_code_end_time_and_provenance_into_the_store() {
    let temp = TempDir::new().unwrap();
    let store = test_store(temp.path());
    let log_path = write_log_with_footer(temp.path(), "2026-09-15 22:21:46.740", 137);
    let record = detached_docker_record(&log_path);
    store.save(&record).unwrap();

    let outcome = finalize_detached_execution(
        &store,
        &record.uuid,
        &DetachedFinalizeFacts {
            exit_code: "137".to_string(),
            oom_killed: "false".to_string(),
            started_at: CONTAINER_STARTED_AT.to_string(),
            finished_at: CONTAINER_FINISHED_AT.to_string(),
            container_error: String::new(),
        },
    );
    assert!(outcome.updated, "reason: {}", outcome.reason);

    let stored = store.get(&record.uuid).unwrap();
    assert_eq!(stored.status, ExecutionStatus::Executed);
    assert_eq!(stored.exit_code, Some(137));
    assert_eq!(stored.oom_killed, Some(false));
    assert_eq!(
        stored.end_time_source.as_deref(),
        Some("docker-finished-at")
    );
    let end_time = DateTime::parse_from_rfc3339(stored.end_time.as_deref().unwrap()).unwrap();
    assert_eq!(
        end_time
            .with_timezone(&Utc)
            .to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        "2026-09-15T22:21:46.740Z"
    );
    assert!(stored.exit_reason.unwrap().contains("SIGKILL"));
}

#[test]
fn finalize_falls_back_to_observation_time_when_docker_reports_no_finish_time() {
    let temp = TempDir::new().unwrap();
    let store = test_store(temp.path());
    let log_path = write_log_with_footer(temp.path(), "2026-09-15 22:21:46.740", 0);
    let record = detached_docker_record(&log_path);
    store.save(&record).unwrap();

    finalize_detached_execution(
        &store,
        &record.uuid,
        &DetachedFinalizeFacts {
            exit_code: "0".to_string(),
            oom_killed: "false".to_string(),
            started_at: "unknown".to_string(),
            // The docker zero-time sentinel must never reach a record.
            finished_at: "0001-01-01T00:00:00Z".to_string(),
            container_error: String::new(),
        },
    );

    let stored = store.get(&record.uuid).unwrap();
    assert_eq!(stored.status, ExecutionStatus::Executed);
    assert_eq!(stored.exit_code, Some(0));
    assert_eq!(stored.end_time_source.as_deref(), Some("observed-at"));
    assert_eq!(stored.observed_at, stored.end_time);
    let end_time = DateTime::parse_from_rfc3339(stored.end_time.as_deref().unwrap()).unwrap();
    assert!(end_time.year() > 2000);
}

#[test]
fn finalize_never_fails_when_the_record_is_gone() {
    let temp = TempDir::new().unwrap();
    let store = test_store(temp.path());
    let outcome = finalize_detached_execution(
        &store,
        "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
        &DetachedFinalizeFacts {
            exit_code: "1".to_string(),
            oom_killed: "false".to_string(),
            started_at: String::new(),
            finished_at: String::new(),
            container_error: String::new(),
        },
    );
    assert!(!outcome.updated);
    assert_eq!(outcome.reason, "record-not-found");
}
